File: code/python/topKFrequent.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        dic1 = {}
        dic2 = {}
        i = 0
        Max = 0
        while i < len(nums):
            if nums[i] not in dic1:
                dic1[nums[i]] = 1
            else:
                dic1[nums[i]] += 1
            i = i + 1
          
        for key in dic1:
            if dic1[key] not in dic2:
                dic2[dic1[key]] = [key]
            else:
                dic2[dic1[key]].append(key) 
            if Max < dic1[key]:
                Max = dic1[key]
        res = []
        while Max > 0:
            if Max in dic2:
                res.extend(dic2[Max])
            if len(res) >=k:
                return res[:k]
            Max -= 1
        
        return res[:k]

File: code/java/topKFrequent.java
class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> map1 = new HashMap<Integer,Integer>();
        List<Integer>[] bucket = new List[nums.length+1];
        List<Integer> res = new ArrayList<Integer>();
        for(int num:nums)
            map1.put(num,map1.getOrDefault(num,0)+1);
        
        for(int key : map1.keySet()){
            int fre = map1.get(key);
            if(bucket[fre] == null)
                bucket[fre] = new ArrayList<Integer>();
            bucket[fre].add(key);
        }
        int i = nums.length;
        int count = 0;
        while(i > 0){
            if(bucket[i] != null)
                res.addAll(bucket[i]);
            if(res.size() >= k)
                break;
            i--;
        }
        return res;
    }
}
